File: ios/GuideStreamTV/Views/AvatarPickerSheet.swift
//
//  AvatarPickerSheet.swift
//  GuideStreamTV
//
//  Lets a signed-in viewer change the picture in their profile header: upload
//  one from the photo library, pick one of the built-in presets, or clear back
//  to the initials monogram.
//
//  Guests get the presets but not the upload. `users.avatar_url` is keyed on
//  the auth uuid and the avatars bucket's RLS policies scope writes to
//  `avatars/<auth.uid()>/…`, so there is nowhere for a guest's file to go and
//  nothing to attach it to — offering the button and failing at the end would
//  be worse than not offering it.
//

import SwiftUI
import PhotosUI

struct AvatarPickerSheet: View {
    @Environment(\.dismiss) private var dismiss
    @State private var auth = AuthViewModel.shared
    @State private var photoItem: PhotosPickerItem?
    @State private var isUploading = false
    @State private var errorText: String?

    let initials: String

    private let columns = [
        GridItem(.flexible(), spacing: 16),
        GridItem(.flexible(), spacing: 16),
        GridItem(.flexible(), spacing: 16),
        GridItem(.flexible(), spacing: 16)
    ]

    var body: some View {
        VStack(spacing: 0) {
            GsSheetHeader(title: "Profile Picture") {
                Button("Done") { dismiss() }
                    .foregroundStyle(Color.textSecondary)
            }

            ScrollView {
                VStack(spacing: 24) {
                    AvatarRing(initials: initials, size: 96, avatar: auth.avatar)
                        .padding(.top, 8)
                        .overlay {
                            if isUploading {
                                ProgressView().tint(.white)
                            }
                        }

                    if auth.isAuthenticated {
                        PhotosPicker(selection: $photoItem, matching: .images, photoLibrary: .shared()) {
                            Label("Upload a photo", systemImage: "photo.on.rectangle.angled")
                                .scaledFont(size: 15, weight: .semibold)
                                .foregroundStyle(.white)
                                .frame(maxWidth: .infinity)
                                .frame(height: 50)
                                .background(
                                    Capsule().fill(Color.orange)
                                )
                        }
                        .disabled(isUploading)
                        .padding(.horizontal, 20)
                    }

                    VStack(alignment: .leading, spacing: 12) {
                        Text("Or pick one")
                            .scaledFont(size: 13, weight: .semibold)
                            .foregroundStyle(Color.textSecondary)
                            .frame(maxWidth: .infinity, alignment: .leading)

                        LazyVGrid(columns: columns, spacing: 16) {
                            ForEach(AvatarPreset.all) { preset in
                                Button {
                                    Task { await auth.setAvatar(preset.storageValue) }
                                } label: {
                                    AvatarRing(
                                        initials: initials,
                                        size: 56,
                                        avatar: .preset(preset.id)
                                    )
                                    .overlay {
                                        if auth.avatarUrl == preset.storageValue {
                                            Circle()
                                                .strokeBorder(Color.orange, lineWidth: 3)
                                                .frame(width: 62, height: 62)
                                        }
                                    }
                                }
                                .buttonStyle(.plain)
                                .accessibilityLabel(preset.id)
                            }
                        }
                    }
                    .padding(.horizontal, 20)

                    Button {
                        Task { await auth.setAvatar(nil) }
                    } label: {
                        Text("Use my initials")
                            .scaledFont(size: 14, weight: .semibold)
                            .foregroundStyle(Color.textSecondary)
                    }
                    .buttonStyle(.plain)

                    if let errorText {
                        Text(errorText)
                            .scaledFont(size: 12)
                            .foregroundStyle(Color.red.opacity(0.9))
                            .multilineTextAlignment(.center)
                            .padding(.horizontal, 20)
                    }

                    Color.clear.frame(height: 24)
                }
            }
        }
        .preferredColorScheme(.dark)
        .sheetSurface(.base)
        .onChange(of: photoItem) { _, item in
            guard let item else { return }
            Task { await upload(item) }
        }
    }

    /// Loads the picked photo, downscales it, and hands the JPEG to the
    /// uploader. Downscaling matters: a modern iPhone photo is several
    /// megabytes and the avatars bucket caps objects at 5 MB, so uploading the
    /// original would fail for exactly the users most likely to try it. 512pt
    /// is comfortably more than the 112pt the largest avatar draws at.
    private func upload(_ item: PhotosPickerItem) async {
        isUploading = true
        errorText = nil
        defer { isUploading = false; photoItem = nil }

        guard let data = try? await item.loadTransferable(type: Data.self),
              let image = UIImage(data: data) else {
            errorText = "That image couldn't be read."
            return
        }
        guard let jpeg = Self.downscaledJPEG(image, maxDimension: 512) else {
            errorText = "That image couldn't be prepared."
            return
        }
        let url = await auth.uploadAvatar(data: jpeg, fileExtension: "jpg", contentType: "image/jpeg")
        if url == nil {
            errorText = "Upload failed. Check your connection and try again."
        }
    }

    /// Aspect-preserving downscale to `maxDimension` on the longest side,
    /// re-encoded as JPEG at 0.85. Returns nil only if rendering fails.
    private static func downscaledJPEG(_ image: UIImage, maxDimension: CGFloat) -> Data? {
        let longest = max(image.size.width, image.size.height)
        let scale = longest > maxDimension ? maxDimension / longest : 1
        let target = CGSize(width: image.size.width * scale, height: image.size.height * scale)
        let renderer = UIGraphicsImageRenderer(size: target)
        let resized = renderer.image { _ in
            image.draw(in: CGRect(origin: .zero, size: target))
        }
        return resized.jpegData(compressionQuality: 0.85)
    }
}

File: android/app/src/main/java/com/rork/guidestreamtvandroid/ui/components/AvatarPickerSheet.kt
package com.rork.guidestreamtvandroid.ui.components

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.net.Uri
import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.result.PickVisualMediaRequest
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.grid.GridCells
import androidx.compose.foundation.lazy.grid.LazyVerticalGrid
import androidx.compose.foundation.lazy.grid.items
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.PhotoLibrary
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.rork.guidestreamtvandroid.data.repository.AuthViewModel
import com.rork.guidestreamtvandroid.ui.theme.BrandOrange
import com.rork.guidestreamtvandroid.ui.theme.SheetLevel
import com.rork.guidestreamtvandroid.ui.theme.SheetSurfaceBase
import com.rork.guidestreamtvandroid.ui.theme.sheetTopInset
import com.rork.guidestreamtvandroid.ui.theme.TextPrimary
import com.rork.guidestreamtvandroid.ui.theme.TextSecondary
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.ByteArrayOutputStream

/**
 * Lets a signed-in viewer change the picture in their profile header: upload
 * one from the photo library, pick one of the built-in presets, or clear back
 * to the initials monogram. Mirrors the iOS AvatarPickerSheet.
 *
 * Guests get the presets but not the upload. `users.avatar_url` is keyed on the
 * auth uuid and the avatars bucket's RLS policies scope writes to
 * `avatars/<auth.uid()>/…`, so there is nowhere for a guest's file to go and
 * nothing to attach it to — offering the button and failing at the end would be
 * worse than not offering it.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun AvatarPickerSheet(
    initials: String,
    onDismiss: () -> Unit,
) {
    val authVm = AuthViewModel.get()
    val context = LocalContext.current
    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)
    val avatarUrl by authVm.accountAvatarUrl.collectAsStateWithLifecycle()

    var isUploading by remember { mutableStateOf(false) }
    var errorText by remember { mutableStateOf<String?>(null) }
    var pickedUri by remember { mutableStateOf<Uri?>(null) }

    val picker = rememberLauncherForActivityResult(
        ActivityResultContracts.PickVisualMedia(),
    ) { uri -> pickedUri = uri }

    LaunchedEffect(pickedUri) {
        val uri = pickedUri ?: return@LaunchedEffect
        isUploading = true
        errorText = null
        val bytes = withContext(Dispatchers.IO) {
            // Downscale before upload. A modern phone photo is several
            // megabytes and the avatars bucket caps objects at 5 MB, so
            // sending the original would fail for exactly the people most
            // likely to try it. 512px is comfortably more than the 88dp the
            // largest avatar draws at.
            runCatching {
                context.contentResolver.openInputStream(uri)?.use { input ->
                    val decoded = BitmapFactory.decodeStream(input) ?: return@use null
                    val longest = maxOf(decoded.width, decoded.height).toFloat()
                    val scale = if (longest > 512f) 512f / longest else 1f
                    val resized: Bitmap = Bitmap.createScaledBitmap(
                        decoded,
                        (decoded.width * scale).toInt().coerceAtLeast(1),
                        (decoded.height * scale).toInt().coerceAtLeast(1),
                        true,
                    )
                    ByteArrayOutputStream().use { out ->
                        resized.compress(Bitmap.CompressFormat.JPEG, 85, out)
                        out.toByteArray()
                    }
                }
            }.getOrNull()
        }
        if (bytes == null) {
            errorText = "That image couldn't be read."
        } else if (authVm.uploadAvatar(bytes) == null) {
            errorText = "Upload failed. Check your connection and try again."
        }
        isUploading = false
        pickedUri = null
    }

    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = sheetState,
        containerColor = SheetSurfaceBase,
        scrimColor = Color.Black.copy(alpha = 0.60f),
        tonalElevation = 0.dp,
        dragHandle = { GsSheetDragHandle(level = SheetLevel.Base) },
        contentWindowInsets = { sheetTopInset() },
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = 20.dp)
                .navigationBarsPadding(),
            horizontalAlignment = Alignment.CenterHorizontally,
        ) {
            Text(
                text = "Profile Picture",
                fontSize = 18.sp,
                fontWeight = FontWeight.Bold,
                color = TextPrimary,
                modifier = Modifier.fillMaxWidth(),
                textAlign = TextAlign.Start,
            )
            Spacer(Modifier.height(18.dp))

            Box(contentAlignment = Alignment.Center) {
                UserAvatar(initials = initials, avatarUrl = avatarUrl, size = 96.dp)
                if (isUploading) {
                    CircularProgressIndicator(color = BrandOrange, modifier = Modifier.size(36.dp))
                }
            }

            Spacer(Modifier.height(20.dp))

            if (authVm.isAuthenticated.value) {
                Row(
                    modifier = Modifier
                        .fillMaxWidth()
                        .height(50.dp)
                        .clip(RoundedCornerShape(25.dp))
                        .background(BrandOrange)
                        .clickable(
                            enabled = !isUploading,
                            interactionSource = remember { MutableInteractionSource() },
                            indication = null,
                        ) {
                            picker.launch(
                                PickVisualMediaRequest(ActivityResultContracts.PickVisualMedia.ImageOnly),
                            )
                        },
                    horizontalArrangement = Arrangement.Center,
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    Icon(
                        imageVector = Icons.Filled.PhotoLibrary,
                        contentDescription = null,
                        tint = Color.White,
                        modifier = Modifier.size(18.dp),
                    )
                    Spacer(Modifier.width(8.dp))
                    Text(
                        text = "Upload a photo",
                        fontSize = 15.sp,
                        fontWeight = FontWeight.SemiBold,
                        color = Color.White,
                    )
                }
                Spacer(Modifier.height(22.dp))
            }

            Text(
                text = "Or pick one",
                fontSize = 13.sp,
                fontWeight = FontWeight.SemiBold,
                color = TextSecondary,
                modifier = Modifier.fillMaxWidth(),
                textAlign = TextAlign.Start,
            )
            Spacer(Modifier.height(12.dp))

            LazyVerticalGrid(
                columns = GridCells.Fixed(4),
                horizontalArrangement = Arrangement.spacedBy(16.dp),
                verticalArrangement = Arrangement.spacedBy(16.dp),
                modifier = Modifier.height(150.dp),
            ) {
                items(AvatarPreset.all) { preset ->
                    Box(contentAlignment = Alignment.Center) {
                        UserAvatar(
                            initials = initials,
                            avatarUrl = preset.storageValue,
                            size = 56.dp,
                            modifier = Modifier.clickable(
                                interactionSource = remember { MutableInteractionSource() },
                                indication = null,
                            ) { authVm.setAvatar(preset.storageValue) },
                        )
                        if (avatarUrl == preset.storageValue) {
                            Box(
                                modifier = Modifier
                                    .size(62.dp)
                                    .clip(CircleShape)
                                    .border(3.dp, BrandOrange, CircleShape),
                            )
                        }
                    }
                }
            }

            Spacer(Modifier.height(16.dp))
            Text(
                text = "Use my initials",
                fontSize = 14.sp,
                fontWeight = FontWeight.SemiBold,
                color = TextSecondary,
                modifier = Modifier.clickable(
                    interactionSource = remember { MutableInteractionSource() },
                    indication = null,
                ) { authVm.setAvatar(null) },
            )

            errorText?.let {
                Spacer(Modifier.height(12.dp))
                Text(
                    text = it,
                    fontSize = 12.sp,
                    color = Color(0xFFEF4444),
                    textAlign = TextAlign.Center,
                )
            }

            Spacer(Modifier.height(24.dp))
        }
    }
}
